src/node_nanomsg.cc

```diff
@@ -191,35 +190,40 @@ NAN_METHOD(Err) {
   info.GetReturnValue().Set(Nan::New(nn_strerror(nn_errno())).ToLocalChecked());
 }
 
-typedef struct nanomsg_socket_s {
-  uv_poll_t poll_handle;
-  uv_os_sock_t sockfd;
-  Nan::Callback *callback;
-} nanomsg_socket_t;
+class NanoMsgPollCtx {
+  private:
+  public:
+    uv_poll_t poll_handle;
+    uv_os_sock_t sockfd;
+    Nan::Callback *callback;
+    NanoMsgPollCtx (int s, bool is_sender, Local<v8::Function> cb) {
+      callback = new Nan::Callback(cb);
+      size_t siz = sizeof(uv_os_sock_t);
+      nn_getsockopt(s, NN_SOL_SOCKET, is_sender ? NN_SNDFD : NN_RCVFD, &sockfd,
+          &siz);
+      poll_handle.data = this;
+    }
+    ~NanoMsgPollCtx () {
+      delete callback;
+    }
+};
 
-void NanomsgReadable(uv_poll_t *req, int status, int events) {
+static void NanomsgReadable(uv_poll_t *req, int status, int events) {
   Nan::HandleScope scope;
 
-  nanomsg_socket_t *context;
-  context = reinterpret_cast<nanomsg_socket_t *>(req);
+  NanoMsgPollCtx *context = reinterpret_cast<NanoMsgPollCtx*>(req->data);
```

Comment:
To clarify some more, in C++, the type `void *` is different from any other `T *` in that `void *` is the top type (base type, e.g. `Object` in Java) of all (data) pointers. Any `T *` can be safely converted to `void *` and back. The reason this is quirky is that `void` itself is actually the bottom type, which cannot be converted to anything, nor can anything be converted to it.

To put it differently, `void *` is not a pointer to `void`, it is a pointer to an unknown type.